service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6260,341 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+
+	prev := maps.Clone(ms.pendingSignalRequestedIDs)
+	for _, requestID := range signalRequestedIds {
+		if _, ok := ms.pendingSignalRequestedIDs[requestID]; !ok {
+			ms.AddSignalRequested(requestID)
+		}
+	}
+	for requestID := range prev {
+		if _, ok := ms.pendingSignalRequestedIDs[requestID]; !ok {
+			ms.DeleteSignalRequested(requestID)
+		}
+	}
+}
+
+func applyUpdatesToSubStateMachine[K comparable, V lastUpdatedStateTransitionGetter](
+	ms *MutableStateImpl,
+	pendingInfos map[K]V,
+	updateInfos map[K]V,
+	updatedSubStateMachine map[K]V,
+	isSnapshot bool,
+	deleteFn func(K) error,
+	fn func(current, incoming V),
+) error {
+	if isSnapshot {
+		for key := range pendingInfos {
+			if _, ok := updatedSubStateMachine[key]; !ok {
+				err := deleteFn(key)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	for key, updated := range updatedSubStateMachine {
+		var existing V
+		if existing, ok := pendingInfos[key]; ok {
+			if CompareVersionedTransition(existing.GetLastUpdateVersionedTransition(), updated.GetLastUpdateVersionedTransition()) == 0 {
+				continue
+			}
+			ms.approximateSize -= existing.Size() + int64SizeBytes
+		}
+		val := updated
+		if fn != nil {
+			val = common.CloneProto(updated)
+			fn(existing, val)
+		}
+		pendingInfos[key] = val
+		updateInfos[key] = val
+		ms.approximateSize += val.Size() + int64SizeBytes
+	}
+	return nil
+}
+
+func (ms *MutableStateImpl) applyUpdatesToUpdateInfos(
+	updatedUpdateInfos map[string]*persistencespb.UpdateInfo,
+) {
+	for updateID, ui := range updatedUpdateInfos {
+		if existing, ok := ms.executionInfo.UpdateInfos[updateID]; ok {
+			if CompareVersionedTransition(existing.GetLastUpdateVersionedTransition(), ui.GetLastUpdateVersionedTransition()) == 0 {
+				continue
+			}
+			ms.approximateSize -= existing.Size() + len(updateID)
+		} else {
+			ms.executionInfo.UpdateCount++
+		}
+		ms.executionInfo.UpdateInfos[updateID] = ui
+		ms.approximateSize += ui.Size() + len(updateID)
+	}
+}
+
+func (ms *MutableStateImpl) syncExecutionInfo(current *persistencespb.WorkflowExecutionInfo, incoming *persistencespb.WorkflowExecutionInfo) error {
+	excludedFields := []string{
+		"WorkflowTaskVersion",
+		"WorkflowTaskScheduledEventId",
+		"WorkflowTaskStartedEventId",
+		"WorkflowTaskRequestId",
+		"WorkflowTaskTimeout",
+		"WorkflowTaskAttempt",
+		"WorkflowTaskStartedTime",
+		"WorkflowTaskScheduledTime",
+		"WorkflowTaskOriginalScheduledTime",
+		"WorkflowTaskType",
+		"WorkflowTaskSuggestContinueAsNew",
+		"WorkflowTaskHistorySizeBytes",
+		"WorkflowTaskBuildId",
+		"WorkflowTaskBuildIdRedirectCounter",
+		"VersionHistories",
+		"ExecutionStats",
+		"LastFirstEventTxnId",
+		"ParentClock",
+		"CloseTransferTaskId",
+		"CloseVisibilityTaskId",
+		"CloseVisibilityTaskCompleted",
+		"WorkflowExecutionTimerTaskStatus",
+		"SubStateMachinesByType",
+		"StateMachineTimers",
+		"TaskGenerationShardClockTimestamp",
+	}
+	err := common.MergeProtoExcludingFields(current, incoming, excludedFields...)
```

Comment:
Will this util method return an error if any field got renamed or is not found in the proto? I feel we need that check or it's quite error prone. 

We also need a test to make sure all those fields still exists so if someone renamed a field it will be caught during development time.